src/rpc/rawtransaction.cpp

```diff
@@ -887,6 +889,231 @@ UniValue signrawtransaction(const JSONRPCRequest& request)
+                // Don't insert mempool ancestors
+                if (ancestors.find(hash_tx->first) == ancestors.end()) {
+                    ordered_tx.push_back(tx);
+                }
+                block_tx.push_back(tx);
+                hash_tx = transactions.erase(hash_tx);
+            } else {
+                hash_tx++;
+            }
+        }
+        if (found_independent) {
+            continue;
+        }
+        // We got all the way through without finding an independent tx
+        result.push_back(Pair("valid",  false));
+        result.push_back(Pair("reason", "circular transaction dependency"));
+        return result;
+    }
+
+    // Should we check against our local policy, or just consensus rules?
+    if (use_local_policy) {
+        // For each input transaction
+        for (auto const& tx : ordered_tx) {
+            const uint256& hash = tx->GetHash();
+            CCoinsViewCache& view = *pcoinsTip;
+
+            // Fail fast if we have an unspent output from this transaction
+            // already in our wallet, which indicates right away that this is a
+            // duplicate tx (copied from sendrawtransaction)
+            bool have_chain = false;
+            for (size_t o = 0; !have_chain && o < tx->vout.size(); o++) {
+                const Coin& existing_coin = view.AccessCoin(COutPoint(hash, o));
+                have_chain = !existing_coin.IsSpent();
+            }
+
+            // Check if transaction already exists in mempool
+            bool have_mempool = mempool.exists(hash);
+
+            if (have_chain) {
+                result.push_back(Pair("valid",  false));
+                result.push_back(Pair("reason", "transaction already in chain"));
+                return result;
+            }
+
+            if (have_mempool) {
+                result.push_back(Pair("valid",  false));
+                result.push_back(Pair("reason", "transaction already in mempool"));
+                return result;
+            }
+
+            // If we are reasonably confident that this transaction is not a
+            // duplicate, do an AcceptToMemoryPool dry run
+            CValidationState state;
+            bool missing_inputs;
+
+            // Call AcceptToMemoryPool with the dryrun flag set
+            if (!AcceptToMemoryPool(mempool, state, std::move(tx), true, &missing_inputs, nullptr, false, true, 0)) {
+                if (state.IsInvalid()) {
+                    result.push_back(Pair("valid",  false));
+                    result.push_back(Pair("reason", strprintf("%i: %s", state.GetRejectCode(), state.GetRejectReason())));
+                    return result;
+                } else if (missing_inputs) {
+                    // We want to ignore the "missing inputs" case here, since the input
+                    // may actually be in the mempool. If the input is truly missing, we
+                    // will find out later after the ancestor and consensus check.
```

Comment:
If ATMP fails for a transaction due to missing inputs, then almost none of the policy checks will be evaluated for the transaction, because that's one of the first things we (must) do -- for instance you can't check the feerate of the transaction without having the inputs available.

Reply:
Actually on further review, it looks like you've added these "dry run" transactions to the mempool, so now I think if ATMP returns "missing inputs", that should reflect an actual failure?  Is there a case where that doesn't hold?

Reply:
Yeah -- this code is leftover from before I figured out the issue you mentioned in your first comment. I was working on getting rid of this when you left your review 😝